File: 100-199/143.py
from math import prod
from euler import SqrtNumber
from collections import defaultdict
from itertools import product
from tqdm import tqdm


OPTS = defaultdict(list)
MAX_RES = 120_000
SEARCH_LIMIT = 4 * MAX_RES


def is_int(x: float) -> bool:
    return x == int(x)


def check_solution(p: int, b: int, c:int) -> int | None:
    q: float = (p + (4 * b ** 2 - 3 * p ** 2) ** 0.5) / 2
    r: float = (p + (4 * c ** 2 - 3 * p ** 2) ** 0.5) / 2
    a: float = (q ** 2 + r ** 2 - q * r) ** 0.5

    if sum(map(is_int, [q,r,a])) == 3:
        print(p,q,r,a,b,c)
        return int(p + q + r)
    return None



def main():
    # inv = SqrtNumber(7, 4 ,3)
    # for n in range(1, SEARCH_LIMIT + 1):
    #     x = SqrtNumber(n, 0, 3)
    #     while x.a <= SEARCH_LIMIT:
    #         x *= inv
    #         if x.a % 2 == 0:
    #             OPTS[x.b].append(x.a // 2)

    for b in tqdm(range(1, 2 * MAX_RES)):
        p = 1
        while 4 * b * b - 3 * p * p >= 0:
            if is_int((4 * b * b - 3 * p * p) ** 0.5):
                OPTS[p].append(b)
            p += 1

    for q in range(1000):
        if OPTS[q]:
            print(q, OPTS[q])

    res = set()

    for p in range(SEARCH_LIMIT):
        for b, c in product(OPTS[p], repeat=2):
            sol = check_solution(p, b, c)
            # print(sol)
            # input()
            if sol:
                res.add(sol)
    
    print(sum([x for x in res if x <= MAX_RES]))


if __name__ == "__main__":
    main()

File: 100-199/143.cpp
#include <iostream>
#include <vector>
#include <set>
#include <cmath>

using namespace std;

typedef long long ll;

ll MAX_RES = 120000;
vector<vector<int>> OPTS(MAX_RES + 1);

bool is_int(double x){
    return (int)x == x;
}

ll check_solution(int p, int b, int c){
    double q = (p + sqrt(4 * b * b - 3 * p * p)) / 2;
    double r = (p + sqrt(4 * c * c - 3 * p * p)) / 2;
    double a = sqrt(q * q + r * r - q * r);
    if(is_int(q) && is_int(r) && is_int(a)){
        return (ll)(p + q + r);
    }
    return 0;
}

int main(){
    for(ll b = 0; b < 2 * MAX_RES; b++){
        ll p = 1;
        while(p < MAX_RES && 4 * b * b - 3 * p * p > 0){
            if(is_int(sqrt(4 * b * b - 3 * p * p))){
                OPTS[p].push_back((b));
            }
            ++p;
        }
    }

    for(ll p = 0; p < 100; p++){
        if(OPTS[p].empty())
            continue;
        cout<<p<<endl;

    }

    set<int> res;
    for(ll p=0; p < MAX_RES; p++){
        for(ll b : OPTS[p]){
            for(ll c : OPTS[p]){
                ll sol = check_solution(p, b, c);
                if(sol && sol <= MAX_RES)
                    res.insert(sol);
            }
        }
    }
    ll sum = 0;
    for(ll x: res){
        sum += x;
    }
    cout<<sum<<endl;
}
